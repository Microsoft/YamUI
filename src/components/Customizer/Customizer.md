### Examples

Button with theme:

```js { "props": { "data-description": "basic" } }
const { createTheme, getTheme } = require('./');
const Button = require('../Button').default;

const grayTheme = createTheme({
  palette: {
    themeDark: '#333',
    neutralDark: '#666',
    themeLighter: '#999',
  },
  fonts: {
    smallPlus: {
      fontSize: '1.8rem',
      lineHeight: '2.4rem',
    },
  }
});

<div>
  <Button text="Normal" />{' '}
  <Customizer settings={{ theme: grayTheme }}>
    <Button text="Big Gray Theme" />
  </Customizer>
</div>;
```
