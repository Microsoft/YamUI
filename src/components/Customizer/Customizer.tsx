/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { Customizer as FabricCustomizer, ICustomizerProps } from 'office-ui-fabric-react/lib/Utilities';
import CustomizableBody from './CustomizableBody';

/**
 * Used to inject a custom theme object into customizable components, overriding the global theme.
 */
export default class Customizer extends React.Component<ICustomizerProps & NestableBaseComponentProps> {
  public render() {
    const { children, className, ...customizerProps } = this.props;

    return (
      <FabricCustomizer {...customizerProps}>
        <CustomizableBody {...{ children, className }} />
      </FabricCustomizer>
    );
  }
}
