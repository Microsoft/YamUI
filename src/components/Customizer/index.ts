/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default, createTheme } from './Customizer';
export * from './Customizer';
