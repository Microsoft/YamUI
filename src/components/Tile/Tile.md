### Notes for use

`Tile` is a container with a certain aspect ratio.

### Examples

With images rendered with different aspect ratios:

```js { "props": { "data-description": "with images" } }
const { TileAspectRatio } = require('.');
const { ImageFit } = require('../Image');

const yammerLogo = 'logo.png';
const yammerLogoDescription = 'Yammer "y" logo';

const image = <Image source={yammerLogo} description={yammerLogoDescription} fullWidth={true} imageFit={ImageFit.cover} />;

<div>
  <FixedGridRow>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
  </FixedGridRow>

  <br />

  <FixedGridRow>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.SixteenNine}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.SixteenNine}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.SixteenNine}>{image}</Tile>
    </FixedGridColumn>
  </FixedGridRow>

  <br />

  <FixedGridRow>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.TwentyOneNine}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.TwentyOneNine}>{image}</Tile>
    </FixedGridColumn>
  </FixedGridRow>
</div>
```

With text:

```js { "props": { "data-description": "with text" } }
const { TileAspectRatio } = require('.');
const { ImageFit } = require('../Image');
const { TextColor } = require('../Text');

const yammerLogo = 'logo.png';
const yammerLogoDescription = 'Yammer "y" logo';

const image = <Image source={yammerLogo} description={yammerLogoDescription} fullWidth={true} imageFit={ImageFit.cover} />;

<div>
  <FixedGridRow>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile aspectRatio={TileAspectRatio.FourThree}>
        <Clickable unstyled={true} block={true} onClick={action('button clicked')}>
          <Text color={TextColor.METADATA}>+ 2 MORE</Text>
        </Clickable>
      </Tile>
    </FixedGridColumn>
  </FixedGridRow>
</div>
```

With border color:

```js { "props": { "data-description": "with border color" } }
const { TileAspectRatio } = require('.');
const { ImageFit } = require('../Image');
const { TextColor } = require('../Text');

const yammerLogo = 'logo.png';
const yammerLogoDescription = 'Yammer "y" logo';

const image = <Image source={yammerLogo} description={yammerLogoDescription} fullWidth={true} />;

<div>
  <FixedGridRow>
    <FixedGridColumn>
      <Tile borderColor="themeDarker" aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile borderColor="themeDarker" aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile borderColor="themeDarker" aspectRatio={TileAspectRatio.FourThree}>{image}</Tile>
    </FixedGridColumn>
    <FixedGridColumn>
      <Tile borderColor="themeDarker" aspectRatio={TileAspectRatio.FourThree}>
        <Clickable unstyled={true} block={true} onClick={action('button clicked')}>
          <Text color={TextColor.METADATA}>+ 2 MORE</Text>
        </Clickable>
      </Tile>
    </FixedGridColumn>
  </FixedGridRow>
</div>
```