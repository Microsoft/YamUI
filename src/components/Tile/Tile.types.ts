/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { PaletteColor } from '../../util/colors';

export { PaletteColor };

export enum TileAspectRatio {
  /** 21:9 aspect ratio */
  TwentyOneNine = '21_9',
  /** 16:9 aspect ratio */
  SixteenNine = '16_9',
  /** 4:3 aspect ratio */
  FourThree = '4_3',
  /** 1:1 aspect ratio */
  OneOne = '1_1',
  /** 2:3 aspect ratio */
  TwoThree = '2_3',
}

export interface TileProps extends NestableBaseComponentProps {
  /**
   * Tile aspect ratio.
   */
  aspectRatio: TileAspectRatio;

  /**
   * If specified, adds a solid 1 px border with the specified theme color to the tile.
   * If not, no border is added.
   */
  borderColor?: PaletteColor;
}
