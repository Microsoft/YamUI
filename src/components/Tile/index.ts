/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './Tile';
export * from './Tile.types';
