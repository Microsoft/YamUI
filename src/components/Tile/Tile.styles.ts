/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { TileAspectRatio } from './Tile.types';
import { memoizeFunction } from 'office-ui-fabric-react/lib/Utilities';
import { mergeStyleSets, ITheme } from 'office-ui-fabric-react/lib/Styling';
import { PaletteColor } from '../../util/colors';

export interface TileClassNameProps {
  aspectRatio: TileAspectRatio;
  borderColor?: PaletteColor;
  theme: ITheme;
}

const mapAspectRatioToPercent: Record<TileAspectRatio, string> = {
  [TileAspectRatio.OneOne]: '100%',
  [TileAspectRatio.FourThree]: '75%',
  [TileAspectRatio.SixteenNine]: '56.25%',
  [TileAspectRatio.TwentyOneNine]: '42.85%',
};

export const getClassNames = memoizeFunction((classNameProps: TileClassNameProps) => {
  const { aspectRatio, borderColor, theme } = classNameProps;

  const borderWidth = borderColor ? '1px' : '0';
  const border = borderColor ? `solid ${borderWidth} ${theme.palette[borderColor]}` : undefined;

  return mergeStyleSets({
    root: {
      paddingBottom: mapAspectRatioToPercent[aspectRatio],
      position: 'relative',
      width: '100%',
      selectors: {
        '&.y-tile > *': {
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          border,
          width: '100%',
          position: 'absolute',
          top: borderWidth,
          bottom: borderWidth,
          left: borderWidth,
          right: borderWidth,
        },
      },
    },
  });
});
