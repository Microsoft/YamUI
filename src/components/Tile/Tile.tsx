/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { CustomizableComponentProps, defaultTheme, customizable } from '../Customizer';
import { join } from '../../util/classNames';
import { getClassNames } from './Tile.styles';
import { TileProps } from './Tile.types';

/**
 * A `Tile` component is a container what maintains a certain aspect ratio.
 */
export class Tile extends React.Component<TileProps & CustomizableComponentProps> {
  public render() {
    const { children } = this.props;
    const classes = this.getClasses();

    return <div className={classes}>{children}</div>;
  }

  private getClasses() {
    const { aspectRatio, className, borderColor, theme = defaultTheme } = this.props;
    return join(['y-tile', className, getClassNames({ aspectRatio, borderColor, theme }).root]);
  }
}

@customizable('Tile', ['theme'])
export default class CustomizableTile extends Tile {}
