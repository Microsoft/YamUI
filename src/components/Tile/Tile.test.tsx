/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import CustomizableTile, { Tile } from './Tile';
import { TileAspectRatio, TileProps } from './Tile.types';
import Customizer, { defaultTheme } from '../Customizer';

describe('<Tile />', () => {
  let component: ShallowWrapper<TileProps>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<Tile aspectRatio={TileAspectRatio.FourThree}>test content</Tile>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(
        <Tile className="TEST_CLASSNAME" aspectRatio={TileAspectRatio.SixteenNine}>
          test content
        </Tile>,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with customizer', () => {
    let mountedComponent: ReactWrapper;
    const theme = defaultTheme;

    beforeEach(() => {
      mountedComponent = mount(
        <Customizer settings={{ theme }}>
          <CustomizableTile aspectRatio={TileAspectRatio.TwentyOneNine} />
        </Customizer>,
      );
    });

    it('receives custom theme', () => {
      expect(mountedComponent.find('CustomizableTile').prop('theme')).toBe(theme);
    });
  });
});
