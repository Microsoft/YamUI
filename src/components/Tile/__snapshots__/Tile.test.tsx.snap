// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> with additional className matches its snapshot 1`] = `
<div
  className=
      y-tile
      TEST_CLASSNAME
      {
        padding-bottom: 56.25%;
        position: relative;
        width: 100%;
      }
      &.y-tile > * {
        align-items: center;
        bottom: 0;
        display: flex;
        justify-content: center;
        left: 0;
        position: absolute;
        right: 0;
        top: 0;
        width: 100%;
      }
>
  test content
</div>
`;

exports[`<Tile /> with borderColor matches its snapshot 1`] = `
<div
  className=
      y-tile
      {
        padding-bottom: 56.25%;
        position: relative;
        width: 100%;
      }
      &.y-tile > * {
        align-items: center;
        border: solid 1px #6c98d9;
        bottom: 1px;
        display: flex;
        justify-content: center;
        left: 1px;
        position: absolute;
        right: 1px;
        top: 1px;
        width: 100%;
      }
>
  test content
</div>
`;

exports[`<Tile /> with default options matches its snapshot 1`] = `
<div
  className=
      y-tile
      {
        padding-bottom: 75%;
        position: relative;
        width: 100%;
      }
      &.y-tile > * {
        align-items: center;
        bottom: 0;
        display: flex;
        justify-content: center;
        left: 0;
        position: absolute;
        right: 0;
        top: 0;
        width: 100%;
      }
>
  test content
</div>
`;
